### PR TITLE
core/laws/testing on scala 3.0.0-M2

### PR DESCRIPTION
**Pull request by hamnis**

### CODE REVIEW

File: core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    validate(ipv4, strCtxExpr, argsExpr)
+  def validateIpv6(strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes) =
+    validate(ipv6, strCtxExpr, argsExpr)
+  def validateMediatype(strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes) =
+    validate(mediatype, strCtxExpr, argsExpr)
+  def validateQvalue(strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes) =
+    validate(qvalue, strCtxExpr, argsExpr)
+
+  def validate[A](validator: Validator[A], strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes): Expr[A] = {
+    strCtxExpr.unlift match {
```
Comment (isomarcte) [inline core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala:57]: You can use `unliftOrError` here and avoid the pattern match (valueOrError in 3.0.0-M3)

File: core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    inline def ipv6(args: Any*): Uri.Ipv6Address = ${LiteralsSyntax.validateIpv6('{ctx}, '{args})}
+    inline def mediaType(args: Any*): MediaType = ${LiteralsSyntax.validateMediatype('{ctx}, '{args})}
+    inline def qValue(args: Any*): QValue = ${LiteralsSyntax.validateQvalue('{ctx}, '{args})}
+  }
+}
+
+object LiteralsSyntax {
+
+  trait Validator[A] {
+    def validate(s: String): Option[String]
```
Comment (isomarcte) [inline core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala:37]: Would this be more clear as `Either[String, Expr[A]]`?

File: core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+package syntax
+
+import scala.quoted._
+
+trait LiteralsSyntax {
```
Comment (isomarcte) [inline core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala:22]: Is this a `trait` so we can extend it in the package object?

**Comment by isomarcte** [comment]
Looks great! I really enjoy how much cleaner Scala 3 macros.

File: core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala (hunk 3 above)
Comment (hamnis) [reply core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala:22]: yes, we need this trait, as the scala source dir references it. It can however be empty. 
It just needs to exist. Since we can have top level extensions, we dont really need this anymore I guess.

Comment (isomarcte) [reply core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala:22]: It might be nice to make the various macro functions, e.g. `validateUri` etc, `private` so we don't expose them in the public API. I noticed that Scala3 doesn't let you call private methods on a companion object in a macro implementation (I'm unclear why), but if we lift them out of the trait, effectively making them private to the package object `literals` we should be able to hide them from the public API.

This could pose problems since we actually export these twice, e.g.`all`/`literals`, but I _think_ we can use the new `export` feature to `export` them into the `all` object, though I've not tested that.

File: core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    validate(s => Uri.Ipv4Address.fromString(s).map(value => Expr(value)), strCtxExpr, argsExpr)
+  def validateIpv6(strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes) =
+    validate(s => Uri.Ipv6Address.fromString(s).map(value => Expr(value)), strCtxExpr, argsExpr)
+  def validateMediatype(strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes) =
+    validate(s => MediaType.parse(s).map(value => Expr(value)), strCtxExpr, argsExpr)
+  def validateQvalue(strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes) =
+    validate(qvalue, strCtxExpr, argsExpr)
+
+  def validate[A](validator: Validator[A], strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes): Expr[A] = {
+    val sc = strCtxExpr.unliftOrError
```
Comment (hamnis) [inline core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala:56]: `strCtxExpr.valueOrError` on M3

File: core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+          ???
+        case Right(expr) => expr
+      }
+    } else {
+      report.error("interpolation not supported", argsExpr)
+      ???
+    }
+  }
+
+  object qvalue extends Validator[QValue] {
```
Comment (hamnis) [inline core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala:75]: I cant get this to compile for some reason. Might this have something to do with private constructor and `AnyVal` ?

File: core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala (hunk 3 above)
Comment (rossabaker) [reply core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala:22]: In log4s, we used top-level definitions in `scala-3`.  It's the only place we _can_ fully embrace Scala 3.

File: core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala (hunk 4 above)
Comment (rossabaker) [reply core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala:56]: I'm not sure we're going to succeed in making it to M2.  Should we just proceed to M3?

**Comment by rossabaker** [comment]
Retargeting after merge/deletion of dotty-main.

**Comment by isomarcte** [comment]
@rossabaker @hamnis Sorry for the probably silly question, I think I'm just out of the loop on how we are handling the Scala 3 builds. How are we building these PRs for testing Scala 3? I don't see any configuration settings in our sbt to enable 3.0.0-M2 and our dependency tree for `core` doesn't have 3.0.0-M2 releases for everything. 

Are people just local publishing the dependencies until things are more stabilized and locally changing the crossScalaVersions?

**Comment by rossabaker** [comment]
@isomarcte That's right. We were cross-building on the `dotty` branch, but were merging a lot of things that broke all Scala versions.  Now we're enforcing that the 2.x builds all still work.  We'll still have to figure out something between when core works on 3.x and all the leaf modules being ready.  Perhaps a separate workflow with a limited project list.  But this is easy and adequate for now.

tl;dr To chip away at Scala 3, `++3.0.0-M2 core/compile` right now.

**Review by rossabaker — APPROVED** [review]
It looks good to me?

**Comment by hamnis** [comment]
I still need to figure out why I cant get the qValue macro to compile, and I'm not entirely sure that the rest compiles either. Removal of the last parboiled stuff will help here.

**Comment by rossabaker** [comment]
Oh, so this doesn't compile on 3.0 yet.  I was misled by the green check.

File: tests/src/test/scala/org/http4s/parser/UriParserSpec.scala [hunk 6]
```diff
@@ -41,7 +33,7 @@ class UriParserSpec extends Http4sSpec {
 
     // http://www.ietf.org/rfc/rfc2396.txt
 
-    "parse a IPv6 address" in {
+    "parse a IPv6 address" in pending {
```
Comment (hamnis) [inline tests/src/test/scala/org/http4s/parser/UriParserSpec.scala:36]: not sure why this does not parse.

Is the generator sound?

**Comment by hamnis** [comment]
This now does multiple things, Maybe it should be split up

File: tests/src/test/scala/org/http4s/parser/UriParserSpec.scala (hunk 6 above)
Comment (rossabaker) [reply tests/src/test/scala/org/http4s/parser/UriParserSpec.scala:36]: I wondered if maybe the new parser was requiring brackets, but it doesn't look like that's it. :thinking:

File: core/src/main/scala/org/http4s/StaticFile.scala [hunk 7]
```diff
@@ -49,8 +49,9 @@ object StaticFile {
       classloader: Option[ClassLoader] = None): OptionT[F, Response[F]] = {
     val loader = classloader.getOrElse(getClass.getClassLoader)
 
+    val acceptEncoding: Option[`Accept-Encoding`] = req.flatMap(_.headers.get(`Accept-Encoding`))
     val tryGzipped =
-      preferGzipped && req.flatMap(_.headers.get(`Accept-Encoding`)).exists { acceptEncoding =>
+      preferGzipped && acceptEncoding.exists { acceptEncoding =>
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/StaticFile.scala:54]: The shadowing of the `acceptEncoding` name messed with my head for a moment.

File: core/src/main/scala/org/http4s/Uri.scala [hunk 8]
```diff
@@ -734,7 +701,8 @@ object Uri {
     *
     * @see https://www.ietf.org/rfc/rfc3986.txt#section-3.21.
     */
-  final case class UserInfo private (username: String, password: Option[String])
+  //todo: rehide constructor when we remove parboiled
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/Uri.scala:704]: Can we move on this now?

File: core/src/main/scala/org/http4s/Uri.scala [hunk 9]
```diff
@@ -1169,7 +1069,10 @@ object Uri {
             Ipv6Address(a, b, c, d, e, f, g, h)
         }
 
-      def repN[A](n: Int, p: cats.parse.Parser0[A], sep: cats.parse.Parser0[Unit] = P.unit): cats.parse.Parser0[List[A]] =
+      def repN[A](
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/Uri.scala:1072]: There's now a `repExactlyAs` in cats-parse that I think could replace this with concision and efficiency, but it doesn't have to be in this PR.

**Review by rossabaker — COMMENTED** [review]
Tremendous work.

File: core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    validate(validator, sc.parts, argsExpr)
+  }
+
+  private def validate[A](validator: Validator[A], parts: Seq[String], argsExpr: Expr[Seq[Any]])(using Type[A])(using Quotes): Expr[A] = {
+    if (parts.size == 1) {
+      val literal = parts.head
+      validator.validate(literal) match {
+        case Some(err) =>
+          report.error(err.message)
+          ???
```
[the remark this answers is not in the document]
Comment (hamnis) [reply core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala:67]: yes, this is to make it type check, I think however there is a reportThrows which might be better suited just here.

File: core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+      report.error("interpolation not supported", argsExpr)
+      ???
+    }
+  }
+
+  object uri extends Validator[Uri] {
+    override def validate(literal: String): Option[ParseFailure] = Uri.fromString(literal).swap.toOption
+
+    override def construct(literal: String)(using Quotes): Expr[Uri] =
+      '{_root_.org.http4s.Uri.unsafeFromString(${Expr(literal)})}
```
[the remark this answers is not in the document]
Comment (hamnis) [reply core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala:80]: Not sure. I found this in ip4s and just ran with it. Nothing else I have tried has worked

File: core/src/main/scala/org/http4s/Uri.scala (hunk 8 above)
Comment (hamnis) [reply core/src/main/scala/org/http4s/Uri.scala:704]: yes

**Comment by rossabaker** [comment]
This looks great. I'll try to figure out the pending test.

File: tests/src/test/scala/org/http4s/parser/UriParserSpec.scala (hunk 6 above)
Comment (m-sp) [reply tests/src/test/scala/org/http4s/parser/UriParserSpec.scala:36]: `value` is skipping leading 0 so it's not the exact same thing.

to do the same thing as the previous test case does we could switch to
```scala
Ipv6Address.parser.string.parseAll(s) must beRight(s)
```

but it cannot parse everything. It is still failing on the following inputs:

```
::32ba:32ba:32ba:32ba:32ba:32ba:32ba  
01ab::32ba:32ba:32ba:32ba:32ba:32ba                                           
01ab:01ab::32ba:32ba:32ba:32ba:32ba
01ab:01ab:01ab::32ba:32ba:32ba:32ba                                           
01ab:01ab:01ab:01ab::32ba:32ba:32ba                                                                                                                          01ab:01ab:01ab:01ab:01ab::32ba:32ba                                           
01ab:01ab:01ab:01ab:01ab:01ab::32ba                                                                                                                          
01ab:01ab:01ab:01ab:01ab:01ab:01ab::
```

so I think the parser might have a small bug

Comment (m-sp) [reply tests/src/test/scala/org/http4s/parser/UriParserSpec.scala:36]:
```
Index: core/src/main/scala/org/http4s/Uri.scala
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/core/src/main/scala/org/http4s/Uri.scala b/core/src/main/scala/org/http4s/Uri.scala
--- a/core/src/main/scala/org/http4s/Uri.scala	(revision e2d897cda6cfced8759fb0cf87cd7e800a641e14)
+++ b/core/src/main/scala/org/http4s/Uri.scala	(date 1611421411712)
@@ -1195,34 +1195,37 @@
       (repN(6, h16Colon).with1 ~ ls32)
         .map { case (ls: collection.Seq[Short], rs) => toIpv6(ls, rs) }
         .backtrack
-        .orElse((doubleColon *> repN(4, h16Colon, P.unit) ~ ls32)
+        .orElse((doubleColon *> repN(5, h16Colon, P.unit) ~ ls32)
           .map { case (rs: List[Short], rs2) => toIpv6(Seq.empty, rs ++ rs2) })
         .backtrack
-        .orElse(((h16.?.with1 <* doubleColon) ~ h16Colon.repExactlyAs[List[Short]](3).backtrack ~ ls32)
+        .orElse(((h16.?.with1 <* doubleColon) ~ h16Colon.repExactlyAs[List[Short]](4).backtrack ~ ls32)
           .map { case ((ls: Option[Short], rs), rs2) => toIpv6(ls.toSeq, rs ++ rs2) })
         .backtrack
         .orElse(
-          ((repN(2, h16, colon.void).?.with1 <* doubleColon) ~ h16Colon.repExactlyAs[List[Short]](2).backtrack ~ ls32)
+          ((repN(3, h16, colon.void).?.with1 <* doubleColon) ~ h16Colon.repExactlyAs[List[Short]](2).backtrack ~ ls32)
+            .map { case ((ls: Option[List[Short]], rs), rs2) => toIpv6(ls.getOrElse(Seq.empty), rs ++ rs2) })
+        .backtrack
+        .orElse(((repN(2, h16, colon.void).?.with1 <* doubleColon) ~ h16Colon.repExactlyAs[List[Short]](3).backtrack ~ ls32)
             .map { case ((ls: Option[List[Short]], rs), rs2) =>
               toIpv6(ls.getOrElse(Seq.empty), rs ++ rs2)
             })
         .backtrack
-        .orElse(((repN(3, h16, colon.void).?.with1 <* doubleColon) ~ h16Colon ~ ls32)
+        .orElse(((repN(4, h16, colon.void).?.with1 <* doubleColon) ~ h16Colon ~ ls32)
           .map { case ((ls: Option[List[Short]], r0: Short), rs) =>
             toIpv6(ls.getOrElse(Seq.empty), Seq(r0) ++ rs)
           })
         .backtrack
-        .orElse(((repN(4, h16, colon.void).?.with1 <* doubleColon) ~ ls32)
+        .orElse(((repN(5, h16, colon.void).?.with1 <* doubleColon) ~ ls32)
           .map { case (ls: Option[collection.Seq[Short]], rs) =>
             toIpv6(ls.getOrElse(Seq.empty), rs)
           })
         .backtrack
-        .orElse(((repN(5, h16, colon.void).?.with1 <* doubleColon) ~ h16)
+        .orElse(((repN(6, h16, colon.void).?.with1 <* doubleColon) ~ h16)
           .map { case (ls: Option[collection.Seq[Short]], rs: Short) =>
             toIpv6(ls.getOrElse(Seq.empty), Seq(rs))
           })
         .backtrack
-        .orElse((repN(6, h16, colon.void).?.with1 <* doubleColon)
+        .orElse((repN(7, h16, colon.void).?.with1 <* doubleColon)
           .map {(ls: Option[collection.Seq[Short]]) => toIpv6(ls.getOrElse(Seq.empty), Seq.empty)})
         .backtrack
     }
```

makes this test pass but has issues with other tests. but there are a couple off by one errors in the ipv6 parser

**Comment by rossabaker** [comment]
@m-sp is taking the lead on fixing the IPv6 parser.  Let's merge this.
